apache/ambari
# [AMBARI-22875] Replay of topology requests with mpacks

## What changes were proposed in this pull request?

Ambari supports replay of cluster creation request in case the server is restarted before the deployment process begins.  This feature does not yet work for mpacks.  The goal of this change is to implement replay of topology requests based on mpacks.

In addition, a minor fix for blueprint export, which currently fails with NPE.

## How was this patch tested?

Tested replay of deployment on live cluster:

1. Start Ambari server, but not agent
2. Submit blueprint, cluster creation request
3. Restart Ambari server
4. Start Ambari agent

```
21 Jun 2018 07:29:30,209  INFO [ambari-client-thread-25] BlueprintResourceProvider:550 - Creating Blueprint, name=blue
21 Jun 2018 07:29:30,386  INFO [ambari-client-thread-26] ClusterResourceProvider:524 - Creating Cluster 'TEST' based on blueprint 'blue'.
21 Jun 2018 07:29:56,396  INFO [main] AmbariServer:1006 - STARTUP_MESSAGE: Starting AmbariServer.java executable
21 Jun 2018 07:30:46,391  INFO [ambari-client-thread-24] TopologyManager:1035 - TopologyManager.replayRequests: no config with TOPOLOGY_RESOLVED found, adding cluster config request
```

Also tested replay of scale request:

1. Submit scale request
2. Restart Ambari server
3. Start another Ambari agent

```
21 Jun 2018 09:31:10,781  INFO [ambari-client-thread-20] ClusterTopologyImpl:260 - ClusterTopologyImpl.addHostTopology: added host = c7402.ambari.apache.org to host group = node
21 Jun 2018 09:31:10,850  INFO [ambari-client-thread-20] TopologyManager:882 - TopologyManager.processRequest: not all required hosts have been matched, so adding LogicalRequest ID = 6 to outstanding requests
21 Jun 2018 09:32:55,763  INFO [main] AmbariServer:1006 - STARTUP_MESSAGE: Starting AmbariServer.java executable
21 Jun 2018 09:33:41,941  INFO [agent-register-processor-1] LogicalRequest:100 - LogicalRequest.offer: attempting to match a request to a request for a reserved host to hostname = c7402.ambari.apache.org
21 Jun 2018 09:33:41,943  INFO [agent-register-processor-1] TopologyManager:636 - TopologyManager.onHostRegistered: processing accepted host offer for reserved host = c7402.ambari.apache.org
```

Note that both test cases may fail due to [AMBARI-23586](https://issues.apache.org/jira/browse/AMBARI-23586), which is fixed on `trunk`, but hasn't been merged into `branch-feature-AMBARI-14714` yet.

Tested export of cluster in blueprint format.
